crates/harp/src/vector/formatted_vector.rs

```diff
@@ -4,233 +4,243 @@
+            RAWSXP => AtomicVector::Raw(unsafe { RawVector::new_unchecked(vector.sexp) }),
+            LGLSXP => AtomicVector::Logical(unsafe { LogicalVector::new_unchecked(vector.sexp) }),
+            INTSXP => AtomicVector::Integer(unsafe { IntegerVector::new_unchecked(vector.sexp) }),
+            REALSXP => AtomicVector::Numeric(unsafe { NumericVector::new_unchecked(vector.sexp) }),
+            STRSXP => {
+                AtomicVector::Character(unsafe { CharacterVector::new_unchecked(vector.sexp) })
+            },
+            CPLXSXP => AtomicVector::Complex(unsafe { ComplexVector::new_unchecked(vector.sexp) }),
+            _ => {
+                return Err(anyhow!("Unsupported type"));
+            },
+        };
+        Ok(vector)
     }
-}
 
-impl<'a> Iterator for FormattedVectorIter<'a> {
-    type Item = String;
+    fn format_element(&self, index: isize) -> String {
+        // We always use the default options for now as this is only used for the variables pane,
+        // we might want to change that in the future.
+        let options = FormatOptions::default();
+        match self {
+            AtomicVector::Raw(v) => v.format_elt_unchecked(index, Some(&options)),
+            AtomicVector::Logical(v) => v.format_elt_unchecked(index, Some(&options)),
+            AtomicVector::Integer(v) => v.format_elt_unchecked(index, Some(&options)),
+            AtomicVector::Numeric(v) => v.format_elt_unchecked(index, Some(&options)),
+            AtomicVector::Character(v) => v.format_elt_unchecked(index, Some(&options)),
+            AtomicVector::Complex(v) => v.format_elt_unchecked(index, Some(&options)),
+        }
+    }
 
-    fn next(&mut self) -> Option<Self::Item> {
-        if self.index == self.size {
-            None
-        } else {
-            let out = Some(self.formatted.get_unchecked(self.index));
-            self.index = self.index + 1;
-            out
+    fn len(&self) -> usize {
+        unsafe {
+            match self {
+                AtomicVector::Raw(v) => v.len(),
+                AtomicVector::Logical(v) => v.len(),
+                AtomicVector::Integer(v) => v.len(),
+                AtomicVector::Numeric(v) => v.len(),
+                AtomicVector::Character(v) => v.len(),
+                AtomicVector::Complex(v) => v.len(),
+            }
         }
     }
 }
 
-impl FormattedVector {
-    pub fn iter(&self) -> FormattedVectorIter {
-        FormattedVectorIter::new(self)
+pub struct FormattedVectorIter {
+    vector: AtomicVector,
+    indices: Box<dyn Iterator<Item = i64>>,
+}
+
+impl FormattedVectorIter {
+    /// Creates a new iterator over the formatted elements of a vector.
+    /// If `indices` is `None`, the iterator will iterate over all elements of the vector.
+    /// If `indices` is `Some`, the iterator will iterate over the elements at the specified indices.
+    /// The indices must be valid and in bounds as we do no bounds checking.
+    fn new_unchecked(
```

Comment:
This method is missing a Safety comment to explain what is being unchecked exactly and what are the failure modes (panics, R errors?).

Reply:
Added some notices here: https://github.com/posit-dev/ark/pull/646/commits/e862651607dbc04090722b4f691f8f99fbcf6a08